### PR TITLE
Use base::SequenceBound to hold an instance of LedgerDatabase

### PR DESCRIPTION
**Pull request by zenparsing**
Resolves https://github.com/brave/brave-browser/issues/15779

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: vendor/bat-native-ledger/BUILD.gn [hunk 1]
```diff
@@ -90,6 +90,8 @@ source_set("headers") {
     "include/bat/ledger/option_keys.h",
   ]
 
+  deps = [ "//sql:sql" ]
```
Comment (zenparsing) [inline vendor/bat-native-ledger/BUILD.gn:93]: @bridiver Is it OK to add this dep to the "headers" source set?

File: vendor/bat-native-ledger/src/bat/ledger/ledger_database.cc [hunk 2]
```diff
@@ -4,12 +4,349 @@
...
+
+mojom::DBRecordPtr CreateRecord(
+    sql::Statement* statement,
+    const std::vector<mojom::DBCommand::RecordBindingType>& bindings) {
+  auto record = mojom::DBRecord::New();
+  int column = 0;
+
+  if (!statement) {
+    return record;
+  }
```
Comment (emerick) [inline vendor/bat-native-ledger/src/bat/ledger/ledger_database.cc:66]: I know this isn't new code, but is this an expected codepath? Or should we just make this a `DCHECK`?

Comment (zenparsing) [reply vendor/bat-native-ledger/src/bat/ledger/ledger_database.cc:66]: Yes, these should be DCHECKs. I'll update the PR.

**Review by emerick — APPROVED** [review]
LGTM

**Review by kylehickinson — APPROVED** [review]
As mentioned in previous PR, would like to make sure everything works at runtime for iOS before this is merged

Code changes lgtm though

**Review by iefremov — APPROVED** [review]
The rewards service part looks fine, except I don't understand whether it's okay to remove those extra `reset()`s. Everything else also look good to me in general

**Comment by iefremov** [comment]
Hey @zenparsing maybe we could give it another try?

**Comment by zenparsing** [comment]
@iefremov I'll rebase and rerun CI today.

**Comment by iefremov** [comment]
@zenparsing maybe we could finalize this one? We continue to receive crashdumps

**Comment by iefremov** [comment]
hey @zenparsing why this was closed? The fix is still needed

**Pull request closed without merging**